### PR TITLE
Merging #21302

### LINKED ISSUE #20129: InputFormat should allow to read subset of columns.
**Use case**

`File`-like table engine with column oriented format (Native, Parquet, Arrow) under the hood.
For more quick data import and for serverless data processing.

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Speeded up reading subset of columns from File-like table engine with internal file written in column oriented data formats (Parquet, Arrow and ORC)
This closes #20129 Done by @keen-wolf.

### CODE REVIEW

File: src/Storages/StorageFile.cpp [hunk 1]
```diff
@@ -412,6 +414,10 @@ class StorageFileSource : public SourceWithProgress
     std::unique_lock<std::shared_timed_mutex> unique_lock;
 };
 
+bool StorageFile::isColumnOriented() const
```
Comment (alexey-milovidov) [inline src/Storages/StorageFile.cpp:417]: Ok.

But better way is to add "properties" to FormatFactory
similar to AggregateFunctionProperties in AggregateFunctionFactory,
StorageFeatures in StorageFactory.

File: src/Formats/FormatFactory.cpp [hunk 2]
```diff
@@ -412,11 +412,25 @@ void FormatFactory::markOutputFormatSupportsParallelFormatting(const String & na
...
     target = true;
 }
 
 
+void FormatFactory::markFormatAsColumnOriented(const String & name)
+{
+    auto & target = dict[name].is_column_oriented;
+    if (target)
+        throw Exception("FormatFactory: Format " + name + " is already marked as column oriented", ErrorCodes::LOGICAL_ERROR);
+}
```
Comment (alexey-milovidov) [inline src/Formats/FormatFactory.cpp:425]: Have you forget the assignment?

**Comment by alexey-milovidov** [comment]
Also it's possible to add a perf test.

File: src/Formats/FormatFactory.cpp (hunk 2 above)
Comment (nikitamikhaylov) [reply src/Formats/FormatFactory.cpp:426]: Nice.

**Comment by nikitamikhaylov** [comment]
I tried to add a performance test, but selecting only one or several columns is extremely fast (and short queries should last at least 0.2s). So let's merge as is. BTW, 
image
Seems like effective reading of columns is worked only with Parquet data format.

**Comment by alexey-milovidov** [comment]
That's awesome!

**Comment by keen-wolf** [comment]
> I tried to add a performance test, but selecting only one or several columns is extremely fast (and short queries should last at least 0.2s). So let's merge as is. BTW,
> image
> Seems like effective reading of columns is worked only with Parquet data format.

yes, I also found this fact through pervious tests. the column pushing down only improve performance for "Parquet", but has little effect on "Arrow" and "ORC",  totally no effect on the "Native" format. I think it's related to the implementation of read path for different formats. in order to improve it, some internal code change within reading lib may be required.

**Comment by akuzm** [comment]
@nikitamikhaylov you should copy changelog entry here, because the original PR was closed and won't show up in changelog. Also you'll have to mention the author manually in it.

**Comment by nikitamikhaylov** [comment]
> @nikitamikhaylov you should copy changelog entry here, because the original PR was closed and won't show up in changelog. Also you'll have to mention the author manually in it.

@akuzm Done.

**Comment by akuzm** [comment]
> > @nikitamikhaylov you should copy changelog entry here, because the original PR was closed and won't show up in changelog. Also you'll have to mention the author manually in it.
> 
> @akuzm Done.

You forgot the CL category :) I changed it to Improvement.

**Pull request merged**